Data Migrations: allow concurrent group migrations

bashtanov opened the pull request:
Currently migrations involving consumer groups cannot run concurrently even if they affect only distinct groups. Allow this.

This involves some changes in backend and worker data structures. Previously it was assumed a topic or a partition is only involved in no more than one migration, but a consumer group topic partition may be touched by multiple migrations.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.3.x
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes
* none

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

bashtanov committed:
c/data-migrate/worker: allow concurrent group migrations

bashtanov committed:
c/data-migrate/backend/coordination: allow concurrent group migrations

Allow concurrent data migrations involving consumer group. This commit
encapsulates changes to the migration coordination in backend.

bashtanov committed:
c/data-migrate/backend: use type aliases for mrstate iterators

bashtanov committed:
rptest/data_migrations_api: test concurrent group migrations

Run multiple of them concurrently, check data consistency

bashtanov committed:
rptest/data_migrations_api: allow concurrent group migrations

Allow concurrent migrations of distinct groups.
Only reject attempts to migrate the group if it is already migrated.

bashtanov committed:
Revert "c/data_migrated_table: disallow concurrent migrations of consumer groups"

This reverts commit da5286c0a2434a58bd01e96b23935e3218aeb53f.

bashtanov committed:
Revert "c/test/data_migrated_table: create only 1 migration with consumer groups"

This reverts commit 4792597ee092d13c8fc483d13a3688df7ca0ffc2.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--3e472b79-836d-4ca9-b024-f7df03d06a6f-->
<!--TEST RESULTS-->
<!--661e5ff3-56f1-42c8-a7e6-46b7feab595d-->
<!--TEST RESULTS-->
<!--eb50efd9-aa23-47df-9687-976b8e70b585-->
<!--TEST RESULTS-->
<!--da2b0b1e-9953-47e1-a7b1-8d1c8d79376a-->
<!--TEST RESULTS-->
<!--c62e59e6-81ce-47ed-bb89-953280260655-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77294">build#77294</a></summary>


| test_class                          | test_method                | test_arguments                                             | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                   | test_history                                                                                                                                                               |
|:------------------------------------|:---------------------------|:-----------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| MountUnmountIcebergTest             | test_simple_remount        | {"cloud_storage_type": 1}                                  | integration | https://buildkite.com/redpanda/redpanda/builds/77294#019ae6f8-89a7-4b2a-8a70-8d6bf33b3214 | FLAKY         | 13/21    | upstream reliability is '78.61271676300578'. current run reliability is '61.904761904761905'. drift is 16.70795 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=MountUnmountIcebergTest&test_method=test_simple_remount                  |
| PartitionForceReconfigurationTest   | test_basic_reconfiguration | {"acks": 1, "controller_snapshots": true, "restart": true} | integration | https://buildkite.com/redpanda/redpanda/builds/77294#019ae6ea-0dff-46ac-bb12-a8dcd795f2a5 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS                | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionForceReconfigurationTest&test_method=test_basic_reconfiguration |
| WriteCachingFailureInjectionE2ETest | test_crash_all             | {"use_transactions": false}                                | integration | https://buildkite.com/redpanda/redpanda/builds/77294#019ae6f8-89a6-461d-a215-01cf952c5f3c | FLAKY         | 15/21    | upstream reliability is '89.24843423799582'. current run reliability is '71.42857142857143'. drift is 17.81986 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all           |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77763">build#77763</a></summary>


| test_class                    | test_method              | test_arguments                                                            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                           | test_history                                                                                                                                                  |
|:------------------------------|:-------------------------|:--------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------|
| FeaturesMultiNodeUpgradeTest  | test_rollback            | null                                                                      | integration | https://buildkite.com/redpanda/redpanda/builds/77763#019b1206-c15b-4803-98c8-952f12e93b4c | FAIL          | 0/11     | Test FAILS after retries.[Significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-B%3A-Test-fails-with-an-improbably-higher-rate-than-baseline-(run-fails))(baseline=0.4097, p0=0.0001, reject_threshold=0.0100) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=FeaturesMultiNodeUpgradeTest&test_method=test_rollback      |
| FeaturesMultiNodeUpgradeTest  | test_upgrade             | null                                                                      | integration | https://buildkite.com/redpanda/redpanda/builds/77763#019b1206-c15c-4edb-8fec-664c1010becb | FAIL          | 0/11     | Test FAILS after retries.[Significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-B%3A-Test-fails-with-an-improbably-higher-rate-than-baseline-(run-fails))(baseline=0.4089, p0=0.0001, reject_threshold=0.0100) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=FeaturesMultiNodeUpgradeTest&test_method=test_upgrade       |
| FeaturesSingleNodeUpgradeTest | test_upgrade             | null                                                                      | integration | https://buildkite.com/redpanda/redpanda/builds/77763#019b1206-c156-44b0-b840-02bd49a12221 | FAIL          | 0/11     | Test FAILS after retries.[Significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-B%3A-Test-fails-with-an-improbably-higher-rate-than-baseline-(run-fails))(baseline=0.4223, p0=0.0002, reject_threshold=0.0100) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=FeaturesSingleNodeUpgradeTest&test_method=test_upgrade      |
| LicenseEnforcementTest        | test_license_enforcement | {"clean_node_after_recovery": false, "clean_node_before_recovery": false} | integration | https://buildkite.com/redpanda/redpanda/builds/77763#019b1206-c15c-4edb-8fec-664c1010becb | FAIL          | 0/11     | Test FAILS after retries.[Significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-B%3A-Test-fails-with-an-improbably-higher-rate-than-baseline-(run-fails))(baseline=0.4149, p0=0.0002, reject_threshold=0.0100) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=LicenseEnforcementTest&test_method=test_license_enforcement |
| LicenseEnforcementTest        | test_license_enforcement | {"clean_node_after_recovery": false, "clean_node_before_recovery": true}  | integration | https://buildkite.com/redpanda/redpanda/builds/77763#019b1206-c15f-4370-9ab0-f98d6d63b022 | FAIL          | 0/11     | Test FAILS after retries.[Significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-B%3A-Test-fails-with-an-improbably-higher-rate-than-baseline-(run-fails))(baseline=0.4155, p0=0.0002, reject_threshold=0.0100) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=LicenseEnforcementTest&test_method=test_license_enforcement |
| LicenseEnforcementTest        | test_license_enforcement | {"clean_node_after_recovery": true, "clean_node_before_recovery": true}   | integration | https://buildkite.com/redpanda/redpanda/builds/77763#019b1206-c160-4cc2-8940-13bd1297be39 | FAIL          | 0/11     | Test FAILS after retries.[Significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-B%3A-Test-fails-with-an-improbably-higher-rate-than-baseline-(run-fails))(baseline=0.4151, p0=0.0002, reject_threshold=0.0100) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=LicenseEnforcementTest&test_method=test_license_enforcement |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77766">build#77766</a></summary>


| test_class                     | test_method               | test_arguments                                                   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                                       |
|:-------------------------------|:--------------------------|:-----------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| AuditLogTestSchemaRegistryACLs | test_sr_audit_authz       | {"audit_transport_mode": "kclient", "endpoint_name": "PUT_MODE"} | integration | https://buildkite.com/redpanda/redpanda/builds/77766#019b1284-e768-43a2-835e-981b2ca0a707 | FLAKY         | 50/51    | Test PASSES after retries.[Inconclusive result after max retries](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#Case-F%3A-inconclusive-result-after-max-retries-(run-passes)%5BinlineExtension%5D)(baseline=0.0000, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.0000, p1=1.0000, trust_threshold=0.5000)                         | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=AuditLogTestSchemaRegistryACLs&test_method=test_sr_audit_authz   |
| PartitionReassignmentsTest     | test_reassignments_cancel | null                                                             | integration | https://buildkite.com/redpanda/redpanda/builds/77766#019b1284-e769-42e1-885b-242adfc7d686 | FLAKY         | 8/11     | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.1166, p0=0.3285, reject_threshold=0.0100. adj_baseline=0.3106, p1=0.3551, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionReassignmentsTest&test_method=test_reassignments_cancel |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/79490">build#79490</a></summary>


| test_class               | test_method              | test_arguments            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                                    |
|:-------------------------|:-------------------------|:--------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------|
| MountUnmountIcebergTest  | test_simple_remount      | {"cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/79490#019be5f5-9b75-4731-b5a4-1beeb3e8c316 | FLAKY         | 8/11     | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.1743, p0=0.5419, reject_threshold=0.0100. adj_baseline=0.4371, p1=0.1146, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=MountUnmountIcebergTest&test_method=test_simple_remount       |
| NodesDecommissioningTest | test_decommission_status | null                      | integration | https://buildkite.com/redpanda/redpanda/builds/79490#019be5f5-9b75-4731-b5a4-1beeb3e8c316 | FLAKY         | 10/11    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0375, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.1083, p1=0.3179, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommission_status |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/79684">build#79684</a></summary>


| test_class               | test_method              | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                                    |
|:-------------------------|:-------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------|
| NodesDecommissioningTest | test_decommission_status | null             | integration | https://buildkite.com/redpanda/redpanda/builds/79684#019bfeff-0975-4a57-9074-8c67acdb3757 | FLAKY         | 10/11    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0427, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.1228, p1=0.2698, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommission_status |
</details>
<!--/body-->

joe-redpanda commented on src/v/cluster/data_migrated_resources.h:
```diff
@@ -48,12 +48,10 @@ class migrated_resources {
         return _topics.contains(topic);
     }
-    /// Checks if consumer group is already migrated
+    /// Checks if consumer group  is already migrated
```
nit

joe-redpanda commented on src/v/cluster/data_migration_worker.cc:
```diff
@@ -121,15 +153,25 @@ worker::perform_partition_work(model::ntp&& ntp, partition_work&& work) {
+    if (ntp_state_it == _managed_ntps.end()) {
+        return;
+    }
+    auto& mntp_states = std::as_const(ntp_state_it)->second->migration_states;
```
q:
whats the point of as_const here?

bashtanov replied:
It's a mistake, I have fixed it, please have a look. There's no special point, just making things const where possible for easier reasoning and to provide compiler more room for optimisation.

joe-redpanda replied:
Ah, that makes sense. Thanks

joe-redpanda commented on src/v/cluster/data_migration_worker.h:
```diff
@@ -73,39 +73,62 @@ class worker : public ss::peering_sharded_service<worker> {
-        bool is_leader;
-        notification_id_type leadership_subscription;
-
         // `last_requested` or `running` must be set. `running->work` and
```
q:
> `last_requested` or `running` must be set.
"One of" or exclusive?

last_requested then is the next batch of work to be performed for a given partition?

bashtanov replied:
> "One of" or exclusive?

The next statement assumes that they both may be set. So "or" is not exclusive.

> last_requested then is the next batch of work to be performed for a given partition?

Whether next or current, it's the one actually needed to be done. If they are different `running` is something no longer needed, cancelled and dying.

bashtanov replied:
I'll rephrase the comment

joe-redpanda commented:
Can you add some high level overview to the commit comments?

joe-redpanda commented on src/v/cluster/data_migration_worker.h:
```diff
@@ -73,39 +73,64 @@ class worker : public ss::peering_sharded_service<worker> {
+
+    using ntp_state_ptr = std::unique_ptr<ntp_state_t>;
+    using managed_ntps_map_t = chunked_hash_map<model::ntp, ntp_state_ptr>;
+    // using ntp_state_it = ntp_state_t::iterator;
```
nit

joe-redpanda replied:
I wonder if we can ask copilot to complain about commented out code, I've missed many of these in my own changes

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/77763-->
<!--dc93f7bd-ea8e-4b94-844d-679d7a6322eb-->
<!--7f1a2cee-ef00-4068-804d-3e63617fffe3-->
<!--1790905a-a532-4945-9954-c5235657aebd-->
<!--bdb58a1e-bc7e-4e1c-a62e-988a32672cdb-->
<!--74695808-e7ec-464d-92b0-5fdf60bc4430-->
<!--title-->
## Retry command for [Build#77763](https://buildkite.com/redpanda/redpanda/builds/77763)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
skip-redpanda-build
skip-units
skip-rebase
tests/rptest/tests/cluster_features_test.py::FeaturesSingleNodeUpgradeTest.test_upgrade
tests/rptest/tests/cluster_features_test.py::FeaturesMultiNodeUpgradeTest.test_rollback
tests/rptest/tests/license_enforcement_test.py::LicenseEnforcementTest.test_license_enforcement@{"clean_node_after_recovery":false,"clean_node_before_recovery":true}
tests/rptest/tests/license_enforcement_test.py::LicenseEnforcementTest.test_license_enforcement@{"clean_node_after_recovery":true,"clean_node_before_recovery":true}
tests/rptest/tests/license_enforcement_test.py::LicenseEnforcementTest.test_license_enforcement@{"clean_node_after_recovery":false,"clean_node_before_recovery":false}
tests/rptest/tests/cluster_features_test.py::FeaturesMultiNodeUpgradeTest.test_upgrade
```
<!--/body-->

bashtanov commented:
/ci-repeat 1

mmaslankaprv reviewed (commented):
few comments

mmaslankaprv commented on src/v/cluster/data_migration_worker.cc:
```diff
@@ -66,31 +66,53 @@ ss::future<> worker::stop() {
+      });
+}
+
+worker::migration_id_key
+worker::make_migration_id_key(const model::ntp& ntp, id migration_id) {
+    return model::topic_namespace_view(ntp) == model::kafka_consumer_offsets_nt
+             ? std::make_optional(migration_id)
+             : std::nullopt;
```
Can we simplify this by always including migration id in the key and making it required for all ntps. The logic preventing migration of the same NTP should exists in higher level.

bashtanov replied:
> The logic preventing migration of the same NTP should exists in higher level.

I'm not sure if it propagates here well. Some stale work from a long gone migration may still be stopping.

mmaslankaprv replied:
can this run concurrently then ? What about stale work related with consumer groups from other migration

bashtanov replied:
We can now, due to using revision ids. I'd still make non-group works sharing the same NTP override each other as an optimisation.

mmaslankaprv commented on src/v/cluster/data_migration_worker.h:
```diff
@@ -48,7 +48,7 @@ class worker : public ss::peering_sharded_service<worker> {
 
 private:
-    struct ntp_state_t {
+    struct mntp_state_t {
```
consider adding a comment explaining the name, it was obvious for `ntp` now it is not

bashtanov committed:
c/data-migrate/worker: abort stale jobs more aggressively

A running work is not needed neither when it is for a different
migration, nor for a different stage of the current migration.

bashtanov committed:
k/s/group_data_parser: move parse_tx_batch outside of global NS

bashtanov committed:
c/commands: fix comment typo

bashtanov committed:
raft/persisted_stm: fix comment typo

bashtanov committed:
c/data-migrate/table: remember migration create and update revisions

These will be used to maintain correct order of operations when working
on individual partitions.

bashtanov committed:
c/data-migrate/backend/remote-nodes: allow concurrent group migrations

Allow concurrent data migrations involving consumer group. This commit
encapsulates changes to backend logic responsible for tracking partition
work delegated to workers on the local node.

bashtanov committed:
c/data-migrate/worker: pass revision id to group un-/block actions

When deleting a group unblock after deletion, not before.

bashtanov committed:
k/s/consumer-groups: revision-aware group un/-blocking

In consumer group partitions, store not only whether each group is
blocked, but also what was the revision of the controller command that
caused it to un-/block.

Un-/block op to fail if it is associated with a revision older than one
already applied.

bashtanov committed:
k/s/group_manager: allow deleting blocked groups

For data migrations to delete groups after blocking them.

bashtanov committed:
k/s/group_data_parser: log access to blocked group as warning

Such requests may sneak in, it's a user error.

bashtanov committed:
c/data-migrate/backend: pass migration update revision id to worker

In migrations backend, remember migration revision id and pass it to
migrations worker.

Migrations backend processes updates from migrations table
asynchronously, as a migration may affect many topics and partitions.
To have migration revision and definition consistent,
backend needs to store both.

bashtanov committed:
k/s/group_manager: revision-aware group deletion when migrating data

Ignore group deletion request from data migrations if another action of
the same kind has already been performed on the group.

revision-aware group deleting

group-deletion

bashtanov committed:
c/partition_properties_stm: avoid out-of-order update using revision_ids

bashtanov committed:
c/data-migrate/backend: use ssx::async_for_each when processing RPC

Such RPC may involve lots of partitions, so it should be processed
asynchronously to avoid reactor stalls.

bashtanov committed:
k/s/group_data_parser: rename CRTP template parameter to `Impl`

`Base` is a misleading name for a CRTP template parameter, as it is
denotes a child class, not a base class.

bashtanov committed:
c/data-migrate: tidy-up

Fix comments formatting, obey rule of 5, remove unused include

joe-redpanda commented on src/v/container/chunked_hash_map.h:
```diff
@@ -146,3 +146,16 @@ memory_usage_lower_bound(const chunked_hash_map<K, V, Hash, EqualTo>& m) {
 }
+
+template<typename K>
+struct fmt::formatter<chunked_hash_set<K>> {
```
I thought we had one of these in untils/to_string

joe-redpanda replied:
https://github.com/redpanda-data/redpanda/commit/fd5890d0fbe258ebed20379a230b867482e6e03c

bashtanov replied:
that one is for `absl::flat_hash_set`, not for `chunked_hash_set`

joe-redpanda replied:
Ah, oops

joe-redpanda commented on src/v/kafka/server/group_manager.cc:
```diff
@@ -1992,12 +1993,16 @@ ss::future<std::error_code> group_manager::empty_and_delete_groups(
 
 ss::future<chunked_vector<deletable_group_result>> group_manager::delete_groups(
-  chunked_vector<std::pair<model::ntp, group_id>> groups) {
+  chunked_vector<std::pair<model::ntp, group_id>> groups, bool allow_blocked) {
```
nit: bool class. Without the editor open its hard to know what bool parameters are indicating

bashtanov replied:
This boolean parameter is supplied in only on call. I don't think introducing a new type is worth it, unless we have a policy of no naked booleans in functions parameters.

joe-redpanda commented on src/v/kafka/server/group_manager.cc:
```diff
@@ -1963,20 +1963,76 @@ group_manager::describe_partition_producers(const model::ntp& ntp) {
+        }
+    }
+
+    co_await ssx::async_for_each(
```
async parallel operations but no gate on group manager, how are we guaranteeing the lifecycle of the group manager?

bashtanov replied:
Good find!

Yes this is a problem. No it cannot be solved by using the kafka group manager's gate, as kafka group manager is destroyed before controller and migration backend&worker are stopped.

The problem is not unique here, it applies to all components passed to `controller::wire_up_and_start`. They are passed to controller late (not into constructor) because they are constructed after controller, so they are probably destructed before controller.

Out of `offsets_upload_requestor`, `producer_id_recovery_manager`, `offsets_recovery_requestor` and `_data_migrations_group_proxy` only the first one looks like passed correctly (while not nicely). It is a shared_ptr instance stolen from a sharded container stopped and destroyed before controller destruction. So, by the time controller needs to be stopped and destroyed, `offsets_upload_requestor` instance is still alive, but has its gate closed.

I'll fix it in a follow-up PR.

joe-redpanda commented on src/v/cluster/data_migration_worker.h:
```diff
@@ -73,39 +74,62 @@ class worker : public ss::peering_sharded_service<worker> {
+    // must run exclusively
+    using migration_id_key = std::optional<id>;
+    static migration_id_key
+    make_migration_id_key(const model::ntp& ntp, id migration_id);
```
I feel like a lot of concurrent migrations hinges on this trick.

I feel like this function should be highlighted somehow to help future readers understand the nesting structure

bashtanov replied:
I'll add a comment

bashtanov replied:
@joe-redpanda this is the only changes since you approved it. Could you re-thumb please?

joe-redpanda commented on src/v/cluster/data_migration_backend.cc:
```diff
@@ -1444,49 +1456,58 @@ ss::future<> backend::process_delta(cluster::topic_table_ntp_delta&& delta) {
+            continue;
+        }
+        auto& tstate = mrstate.outstanding_topics.at(nt);
+        clear_tstate_belongings({nt, migration_id}, tstate);
```
why is tstate being cleared on a delta without regard for what the delta is?

bashtanov replied:
1. [We roughly know what it is](https://github.com/redpanda-data/redpanda/blob/4d5759c7d6dc9f437193469ccebdbd4020e6bee3/src/v/cluster/data_migration_backend.cc#L1451-L1472)
2. Right below we call `reconcile_existing_topic` to re-init tstate.

joe-redpanda commented:
What are the sources for the revision ids of the block / unblock operations in migrations, is it the record sent when a given migration is created?


A further along controller batch can override a block / unblock on a given partition, at which point migration operations on that partition will be rejected.
What does the UX look like if this happens?
Workers retry forever but afaik can't make any progress. the user will likely notice that their migration is gummed up. Cancel migration & restart?

bashtanov commented:
> What are the sources for the revision ids of the block / unblock operations in migrations, is it the record sent when a given migration is created?

[It's the controller offset of migration creation or update command](https://github.com/redpanda-data/redpanda/pull/28828/commits/f4c503b9bcf85f4ea3721b1bab1dbd50b9d88a3d#diff-eef6604e655d86bbc7100a42af3ddac0857e047cf8641b5d28078113cf5de11fR93). In case of update it's transitioning the migration into a new stage.

> A further along controller batch can override a block / unblock on a given partition, at which point migration operations on that partition will be rejected. What does the UX look like if this happens?

This may only happen when a migration has transitioned into a new stage, or a new migration involving the same entity (topic, revision_id) is created. The latter also indicates that the first migration has transitioned into a new stage, as otherwise the new migration cannot be created.

So the operation with the older revision id must be something that has already succeeded, but worker for some reason id not aware about it. It can be that it succeeded on a different replica of the same partition, or there was a coordination problem (e.g. migration coordinator ~aka raft0 leader went down).

Block/unblocked revisions are saved per entity, i.e. per group in case of __consumer_offsets partition, and per-partition in case of a data partition. If a further revision has been applied then the operations attributed to the earlier revision are not desired to complete. In fact, they are known to have been completed by that time, it's just this particular shard on this particular node only knows this from revision id comparison, as coordination was not quick enough.

The UX will be not observing changes that are part of a long completed migration stage. E.g. a consumer group won't be deleted again after it was migrated away and recreated. Or a partition won't be blocked again after it was blocked for a migration but then unblocked as the migration was cancelled.

> Workers retry forever but afaik can't make any progress. the user will likely notice that their migration is gummed up. Cancel migration & restart?

Workers retry forever unless backend instructs them to stop. An operation stuck in worker because of a revision id mismatch indicates that the migration has already transitioned to a further state. At some point the worker will be informed about it, and the operation will be aborted.

mmaslankaprv commented on src/v/cluster/partition_properties_stm.cc:
```diff
@@ -172,14 +176,28 @@ void partition_properties_stm::apply_record(
-    bool differs = are_writes_disabled() != update.writes_disabled;
+    const auto current_state = _state_snapshots.back();
+    bool is_legacy_command = update.writes_revision_id == model::revision_id{};
+    bool is_newer_revision = update.writes_revision_id
+                             > current_state.writes_revision_id;
```
should that be `>=` ?

bashtanov replied:
a no-op command (where both boolean flag and revisions are the same as in the state) cannot make it through to here, as it's ignored on ingestion in `replicate_properties_update`

a command with the same data but a different value for the boolean flag is deemed invalid

mmaslankaprv commented on src/v/cluster/partition_properties_stm.cc:
```diff
@@ -198,20 +216,57 @@ partition_properties_stm::apply_raft_snapshot(const iobuf& buffer) {
+    }
+    if (
+      cmd.writes_revision_id != model::revision_id{}
+      && cmd.writes_revision_id <= current_state.writes_revision_id)
```
i think we should enable changes if revision is `>=` to the current one

bashtanov replied:
yes, and that's what we are doing: 
* in the block above the case of repeated change is processed 
* in this block we fail if the new revision is less that the current, or if they are equal but the boolean data (disable/enable) is different

joe-redpanda commented:
> It's the controller offset of migration creation or update command

Solid explanation, thanks!

The instance I'm thinking of was if a user sends a block / unblock command on one of these entities concurrent with a migration on it. Do we have an escape hatch that can force unjam / cancel a migration?

bashtanov commented:
Currently only data migrations send these commands to data and consumer groups partitions. If we allow any other component, or the user indeed, to send these, we'll need these to supply a revision id consistent with the migration-originated commands.

However, on the STM/partition API level there is an unintended escape hatch: a legacy revision-less batch is always accepted, and it resets the state so that the next batch is accepted too regardless the revision.

joe-redpanda reviewed (dismissed):
LGTM save abort stuff which it sounds like you're tracking in a follow on
